staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -44,10 +45,34 @@ func newReady() *ready {
 }
 
 // wait blocks until it is Ready or Stopped, it returns an error if is Stopped.
-func (r *ready) wait() error {
+func (r *ready) wait(ctx context.Context) error {
 	r.c.L.Lock()
 	defer r.c.L.Unlock()
+
+	if r.state == Pending {
+		// We're still waiting for initialization.
+		// Ensure that context is honored in case it is cancelled
+		// before initialization happens.
+		// However avoid spanning additional goroutine in a healthy case.
+		// To prevent that this goroutine will be leaked if the context
+		// is never cancelled, force cancel it on exit from the function.
+		waitCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		go func() {
+			<-waitCtx.Done()
+
+			r.c.L.Lock()
+			defer r.c.L.Unlock()
+			r.c.Broadcast()
```

Comment:
the alternative is replace the condition variable by a channel  https://github.com/kubernetes/kubernetes/pull/116024/

At the end , we just need to synchronize everything in this pending state, a channel will do and the lock with the enum will handle the different states

Reply:
the con if this approach is that in an scenarios with a lot of goroutines waiting we are going to send as many `Broadcast` as goroutines , I don't know if this can present a problem